@xen-orchestra/backups-cli/index.js

```diff
@@ -0,0 +1,201 @@
+      }
+    } catch (error) {
+      console.warn('Error while checking VHD', path)
+      console.warn('  ', error)
+      if (error != null && error.code === 'ERR_ASSERTION') {
+        console.warn('  deleting…')
+        console.warn('')
+        force && (await fs.unlink(path))
+      }
+    }
+  })
+
+  // remove VHDs with missing ancestors
+  {
+    const deletions = []
+
+    const check = (parent, child) => {
+      // no longer needs to be checked
+      delete vhdParents[child]
+
+      // check the ancestors first
+      const grandparent = vhdParents[parent]
+      if (grandparent !== undefined) {
+        check(grandparent, parent)
+      }
+
+      if (!vhds.has(parent)) {
+        vhds.delete(child)
+
+        console.warn('Error while checking VHD', child)
+        console.warn('  missing parent', parent)
+        console.warn('  deleting…')
+        console.warn('')
+        force && deletions.push(fs.unlink(child))
+      }
+    }
+
+    // > A property that is deleted before it has been visited will not be
+    // > visited later.
+    // >
+    // > -- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#Deleted_added_or_modified_properties
+    for (const child in vhdParents) {
+      check(vhdParents[child], child)
+    }
+
+    await Promise.all(deletions)
+  }
+
+  const [jsons, xvas] = await readDir(vmDir).then(entries => [
+    entries.filter(endsWith('.json')),
+    entries.filter(endsWith('.xva')),
+  ])
+
+  const unusedVhds = new Set(vhds)
+  const unusedXvas = new Set(xvas)
+
+  await asyncMap(jsons, async json => {
+    const metadata = JSON.parse(await fs.readFile(json))
+    const { mode } = metadata
+    if (mode === 'full') {
+      const linkedXva = resolve(vmDir, metadata.xva)
+
+      if (xvas.has(linkedXva)) {
+        unusedXvas.delete(resolve(vmDir, linkedXva))
+      } else {
+        console.warn('Error while checking backup', json)
+        console.warn('  missing file', linkedXva)
+        console.warn('  deleting…')
+        console.warn('')
+        force && (await fs.unlink(json))
+      }
+    } else if (mode === 'delta') {
+      let linkedVhds = metadata.vhds
+      linkedVhds = Object.keys(linkedVhds).map(key =>
+        resolve(vmDir, linkedVhds[key])
+      )
+
+      if (linkedVhds.every(_ => vhds.has(_))) {
+        linkedVhds.forEach(_ => unusedVhds.delete(_))
+      } else {
+        console.warn('Error while checking backup', json)
+        linkedVhds
+          .filter(_ => !vhds.has(_))
+          .forEach(linkedVhd => {
+            console.warn('  missing file', linkedVhd)
+          })
+        console.warn('  deleting…')
+        console.warn('')
+        force && (await fs.unlink(json))
+      }
+    }
+  })
+
+  await Promise.all([
+    asyncMap(unusedXvas, path => {
+      console.warn('Unused XVA', path)
+      console.warn('  deleting…')
+      console.warn('')
+      return force && fs.unlink(path)
+    }),
+  ])
```

Comment:
What's the point of `Promise.all`?

Reply:
It's only here because I'm working on adding the handling of VHDs in parallel.